[WIP] STAC-24067: use custom integrations for extending the interface (#173)
rancher/stackstate-product-docs: docs/latest/modules/en/pages/setup/custom-integrations/metric-bindings/index.adoc

ifdef::ss-ff-stackpacks2_enabled[]
= Add custom charts to components
:revdate: 2025-07-10
:page-revdate: {revdate}
:description: SUSE Observability

== Overview

> SUSE Observability provides already many metric charts by default on most types of components that represent Kubernetes resources. Extra metric charts can be added to any set of components whenever needed. When adding metrics to components there are 2 options:

Reviewer: ```suggestion
{stackstate-product-name} provides already many metric charts by default on most types of components that represent Kubernetes resources. Extra metric charts can be added to any set of components whenever needed. When adding metrics to components there are two options:
```
Numericals are usually used only for numbers more than 10.

. The metrics are already collected by SUSE Observability but aren't visualized on a component by default
. The metrics aren't yet collected by SUSE Observability at all and therefore aren't available yet

For option 1 the steps below will instruct you how to create a metric binding which will configure SUSE Observability to add a specific metric to a specific set of components.

In case of option 2, first make sure that the metrics are available in SUSE Observability, by sending them to SUSE Observability using the xref:/use/metrics/k8s-prometheus-remote-write.adoc[Prometheus remote write protocol]. Only after that continue by adding charts for the metrics to the components.

== Steps

Steps to create a metric binding:

. <<_create_an_outline_of_the_metric_binding,Create an outline of the metric binding>>
. <<_select_the_components_to_bind_to,Select the components to bind to>>
. <<_write_the_promql_query,Write the PromQL query for the desired metric>>
. <<_bind_the_correct_time_series_to_each_component,Bind the correct time series to each component>>
. <<_create_or_update_the_metric_binding_in_suse_observability,Create or update the metric binding in SUSE Observability>>

> As an example the steps will add a metric binding for the `Replica counts`  of Kubernetes deployments. This is just an example, this metric binding already exists in SUSE Observability by default.

Reviewer: ```suggestion
For example, the steps will add a metric binding for the `Replica counts`  of Kubernetes deployments. This metric binding already exists in {stackstate-product-name}, by default.
```

== Create an outline of the metric binding

> Create a new YAML file called `metricbindings.yaml` and add this YAML template to it to create your own metric binding. If you used the `sts stackpack` command, this file will already be present.  Open it in your favorite code editor to change it throughout this guide. At the end the SUSE Observability CLI will be used to create and update the metric bindings in SUSE Observability.

Reviewer: ```suggestion
Create a new YAML file called `metricbindings.yaml` and add this YAML template to it to create your own metric binding. If you used the `sts stackpack` command, this file will already be present.  Open it in your favorite code editor to change it throughout this guide. At the end, the {stackstate-product-name} CLI will be used to create and update the metric bindings in {stackstate-product-name}.
```

== Select the components to bind to

Save a view of the xref:/use/views/k8s-topology-perspective.adoc[Topology perspective] and use the filters (Filters -> Topology -> Switch to STQL) to query the components that need to show the new metric. The most common fields to select topology on for metric bindings are `type` for the component type and `label` for selecting all the labels. For example for the deployments:

----
type = "deployment" and label = "stackpack:kubernetes"
----

The type filter selects all deployments, while the label filter selects only components created by the Kubernetes stackpack (label name is `stackpack` and label value is `kubernetes`). The latter can also be omitted to get the same result.  All xref:/develop/reference/k8sTs-stql_reference.adoc#_component_filters[STQL query Component Filters] can be used for filtering.

Switch to the advanced mode to copy the resulting topology query and put it in the `scope` field of the metric binding.

[NOTE]
====

> Metric bindings only support the query filters, query functions like `withNeighborsOf` aren't supported and can't be used.

Reviewer: ```suggestion
Metric bindings only support the query filters. Query functions like `withNeighborsOf` are not supported and cannot be used.
```

====

== Write the PromQL query

> Go to the xref:/use/metrics/k8sTs-explore-metrics.adoc[metric explorer] of your SUSE Observability instance, http://your-instance/#/metrics, and use it to query for the metric of interest. The explorer has auto-completion for metrics, labels, label values but also PromQL functions, and operators to help you out. Start with a short time range of, for example, an hour to get the best results.

Reviewer: ```suggestion
Go to the xref:/use/metrics/k8sTs-explore-metrics.adoc[metric explorer] of your {stackstate-product-name} instance, http://your-instance/#/metrics, and use it to query for the metric of interest. The explorer has auto-completion for metrics, labels, label values but also PromQL functions, and operators to help you out. Start with a short time range of, for example, an hour to get the best results.
```

----

image::k8s/k8s-replica-counts-with-binding.png[After adding the parameterized filters the resulting chart looks as expected, only 1 time series for this component]

The PromQL query now filters on 3 labels, `cluster_name`, `namespace` and `deployment`. Instead of specifying an actual value for these labels a variable reference to fields of the component is used. In this case the labels `cluster-name` and `namespace` are used, referenced using `${tags.cluster-name}` and `${tags.namespace}`. Further the component name is referenced with `+${name}+`.

Supported variable references are:

* Any component label, using `${tags.<label-name>}`
* The component name, using `+${name}+`

image::k8s/k8s-carts-highlights.png[Component Highlights page that shows the labels and component name (both highlighted in red)]

[NOTE]
====
The cluster name, namespace and a combination of the component type and name are ususally enough for selecting the metrics for a specific component from Kubernetes. These labels, or similar labels, are usually available on most metrics and components.
====

== Create or update the metric binding in SUSE Observability

> Use the SUSE Observability CLI to create the metric binding in SUSE Observability. Make sure the `metric-bindings.yaml` is saved and looks like this:

Reviewer: ```suggestion
Use the {stackstate-product-name} CLI to create the metric binding in {stackstate-product-name}. Make sure the `metric-bindings.yaml` is saved and looks like this:
```
